File: libtcod_client/server_API.py
import redis
from marshal import dumps
#import sys
#import json

class Admin_commands:
	
	def __init__(self):
		self.r_client = redis.Redis('localhost')

	#sends message to server
	def send_message(self,msg, world_id = 0):
		world_id = msg['world_id']
		self.r_client.lpush("world_"+str(world_id), dumps(msg))

	# required = ['position']
	# optional = ['player_id']
	## command for creating a new agent at position (x,y,z)
	def create_agent(self, x, y, z, player_id = 0):
		#create msg to be sent to server
		msg = {}
		#specify command type and command
		msg['type'] = 'admin'
		msg['cmd'] = 'create_agent'
		#command parameters
		msg['world_id'] = 0
		msg['position'] = (0,x,y,z) #position is a 4, tuple (position_type, x, y, z)
		msg['player_id'] = player_id
		#send message to server
		self.send_message(msg)

	def create_item(self, x, y, z):
		msg = {}
		msg['type'] = 'admin'
		msg['cmd'] = 'create_item'
		msg['world_id'] = 0
		msg['position'] = (0,x,y,z) #position is a 4, tuple (position_type, x, y, z)
		self.send_message(msg)		

	def create_object(self, x, y, z, object_type = None, template = None):
		msg = {}
		msg['type'] = 'admin'
		msg['cmd'] = 'create_object'
		msg['object_type'] = object_type
		msg['template'] = template
		msg['world_id'] = 0
		msg['position'] = (0,x,y,z) #position is a 4, tuple (position_type, x, y, z)
		self.send_message(msg)	
	
	def create_crop(self, x, y, z, template = None):
		self.create_object(x,y,z, 'crop', template)
		
	#required ['position', 'value', 'world_id']
	#optional []		
	def set_map(self, x, y, z, value):
		msg = {}
		msg['type'] = 'admin'
		msg['cmd'] = 'set_map'
		msg['world_id'] = 0
		msg['position'] = (0,x,y,z) #position is a 4, tuple (position_type, x, y, z)
		msg['value'] = value
		self.send_message(msg)			
	
class Info_commands:
	
	def __init__(self):
		self.r_client = redis.Redis('localhost')

	def send_message(self,msg, world_id = 0):
		"""Sends message to redis server."""
		world_id = msg['world_id']
		self.r_client.lpush("world_"+str(world_id), dumps(msg))

	#required = ['z', 'client_id', 'world_id']
	#optional = ['z']
	def get_map(self, world_id=0, z=0):
		msg = {}	
		msg['type'] = 'info'
		msg['cmd'] = 'get_map'
		msg['world_id'] = world_id
		msg['z'] = z
		msg['client_id'] = 0 #will need this
		self.send_message(msg)

	#required = ['agent_id', 'client_id', 'world_id']
	#optional = []			
	def get_agent(self, id):
		msg = {}	
		msg['type'] = 'info'
		msg['cmd'] = 'get_agent'
		msg['world_id'] = 0
		msg['client_id'] = 0
		msg['agent_id'] = id
		self.send_message(msg)

	#required = ['object_id', 'client_id', 'world_id']
	#optional = []			
	def get_object(self, id):
		msg = {}	
		msg['type'] = 'info'
		msg['cmd'] = 'get_object'
		msg['world_id'] = 0
		msg['client_id'] = 0
		msg['object_id'] = id
		self.send_message(msg)		
		
	#required = ['client_id', 'world_id']
	#optional = []
	def get_agent_list(self, world_id = 0):
		msg = {}	
		msg['type'] = 'info'
		msg['cmd'] = 'get_agent_list'
		msg['world_id'] = world_id
		msg['client_id'] = 0 #will need this
		self.send_message(msg)

	#required = ['client_id', 'world_id']
	#optional = []
	def get_object_list(self, world_id = 0):
		msg = {}	
		msg['type'] = 'info'
		msg['cmd'] = 'get_object_list'
		msg['world_id'] = world_id
		msg['client_id'] = 0 #will need this
		self.send_message(msg)		

	#required = ['client_id', 'world_id']
	#optional = []
	def get_tiles(self, world_id = 0):
		msg = {}	
		msg['type'] = 'info'
		msg['cmd'] = 'get_tiles'
		msg['world_id'] = world_id
		msg['client_id'] = 0 #will need this
		self.send_message(msg)		

class Agent_commands:
	
	def __init__(self):
		self.r_client = redis.Redis('localhost')

	def send_message(self,msg, world_id = 0):
		world_id = msg['world_id']
		self.r_client.lpush("world_"+str(world_id), dumps(msg))

	# required = ['dp, agent_id']
	# optional = ['player_id']
	def move_0(self, agent_id, dx, dy, dz):
		msg = {}	
		msg['type'] = 'agent'
		msg['cmd'] = 'move_0'
		msg['world_id'] = 0
		msg['agent_id'] = agent_id
		msg['dp'] = (dx,dy,dz) #position is a 4, tuple (position_type, x, y, z)
		self.send_message(msg)

	# required = ['agent_id', 'target_id']
	# optional = ['target_type', 'attack_type']
	def attack(self, agent_id, target_id, target_type = None, attack_type = None):
		msg = {}	
		msg['type'] = 'agent'
		msg['cmd'] = 'attack'
		msg['agent_id'] = agent_id
		msg['target_id'] = target_id
		msg['world_id'] = 0
		if target_type != None:
			msg['target_type'] = target_type
		if attack_type != None:
			msg['attack_type'] = attack_type
		self.send_message(msg)			
	
	# required = ['position', agent_id']
	# optional = ['action']
	def dig(self, agent_id, x, y, z, action = None):
		msg = {}	
		msg['type'] = 'agent'
		msg['cmd'] = 'dig'
		msg['position'] = (0, x, y, z)
		msg['world_id'] = 0
		msg['agent_id'] = agent_id
		if action != None:
			msg['action'] = action
		self.send_message(msg)		

	#required = ['client_id', 'world_id']
	#optional = []
	def till_soil(self, agent_id):
		msg = {}	
		msg['type'] = 'agent'
		msg['cmd'] = 'till_soil'
		msg['world_id'] = 0
		msg['agent_id'] = agent_id
		self.send_message(msg)		

	#required = ['client_id', 'world_id']
	#optional = []
	def plant_crop(self, agent_id):
		msg = {}	
		msg['type'] = 'agent'
		msg['cmd'] = 'plant_crop'
		msg['world_id'] = 0
		msg['agent_id'] = agent_id
		self.send_message(msg)		
		
	def harvest_crop(self, agent_id, crop_id = 0):
		msg = {}	
		msg['type'] = 'agent'
		msg['cmd'] = 'harvest_crop'
		msg['world_id'] = 0
		msg['agent_id'] = agent_id
		msg['crop_id'] = crop_id
		self.send_message(msg)
	
	def craft_item(self, agent_id, recipe):
		msg = {}
		msg['type'] = 'agent'
		msg['cmd'] = 'craft_item'
		msg['world_id'] = 0
		msg['agent_id'] = agent_id
		msg['recipe'] = recipe
		self.send_message(msg)	

	# required = ['position, agent_id', 'wall_type']
	# optional = []		
	def construct_wall(self, agent_id, x, y, z, wall_type = 'generic_wall'):
		msg = {}
		msg['type'] = 'agent'
		msg['cmd'] = 'construct_wall'
		msg['world_id'] = 0
		msg['agent_id'] = agent_id
		msg['position'] = (0, x, y, z)
		msg['wall_type'] = wall_type
		self.send_message(msg)			

	def pickup_item(self, agent_id, item_id):
		msg = {}
		msg['type'] = 'agent'
		msg['cmd'] = 'pickup_item'
		msg['world_id'] = 0
		msg['agent_id'] = agent_id
		msg['item_id'] = item_id
		self.send_message(msg)			

	def drop_item(self, agent_id, item_id = None):
		msg = {}
		msg['type'] = 'agent'
		msg['cmd'] = 'drop_item'
		msg['world_id'] = 0
		msg['agent_id'] = agent_id
		if item_id != None:
			msg['item_id'] = item_id
		self.send_message(msg)

class Controller_commands:
	
	def __init__(self):
		self.r_client = redis.Redis('localhost')

	def send_message(self,msg, world_id = 0):
		world_id = msg['world_id']
		self.r_client.lpush("world_"+str(world_id), dumps(msg))

	# required = []
	# optional = []
	def move(self, agent_id, x, y, z):
		msg = {}	
		msg['type'] = 'controller'
		msg['cmd'] = 'move_agent_goal'
		msg['world_id'] = 0
		msg['agent_id'] = agent_id
		msg['position'] = (0, x,y,z) #position is a 4, tuple (position_type, x, y, z)
		self.send_message(msg)

	# required = []
	# optional = []
	def move_item(self, agent_id, item_id, x=0, y=0, z=0):
		msg = {}	
		msg['type'] = 'controller'
		msg['cmd'] = 'move_item_goal'
		msg['world_id'] = 0
		msg['agent_id'] = agent_id
		msg['item_id'] = item_id
		msg['position'] = (0, x,y,z) #position is a 4, tuple (position_type, x, y, z)
		self.send_message(msg)
		
class Job_manager_commands:
	
	def __init__(self):
		self.r_client = redis.Redis('localhost')
		
	def send_message(self,msg, world_id = 0):
		world_id = msg['world_id']
		self.r_client.lpush("world_"+str(world_id), dumps(msg))
		
	# required = []
	# optional = []
	def create_job(self, player_id, job):
		msg = {}	
		msg['type'] = 'manager'
		msg['cmd'] = 'create_job'
		msg['world_id'] = 0
		msg['player_id'] = player_id
		msg['job'] = job
		self.send_message(msg)
		
	# required = []
	# optional = []
	def delete_job(self, job_id, player_id = 0):
		msg = {}	
		msg['type'] = 'manager'
		msg['cmd'] = 'delete_job'
		msg['world_id'] = 0
		msg['player_id'] = player_id
		msg['job_id'] =  job_id
		self.send_message(msg)

	def construct_tile(self, tile_id, player_id, build_from, build_at):
		job = {
		'type' : 'construct_tile',
		'tile_id' : tile_id,
		'material_list' : [],
		'build_from' : build_from,
		'build_at' : build_at,	
		}
		self.create_job(player_id, job)
		
	## required = []
	## optional = []
	#def move_item(self, agent_id, item_id, x=0, y=0, z=0):
		#msg = {}	
		#msg['type'] = 'manager'
		#msg['cmd'] = 'move_item_goal'
		#msg['world_id'] = 0
		#msg['agent_id'] = agent_id
		#msg['item_id'] = item_id
		#msg['position'] = (0, x,y,z) #position is a 4, tuple (position_type, x, y, z)
		#self.send_message(msg)	

File: html_client/js2/server_api.js
var globals = {

    world_id: 0,
    client_id: 0,
    player_id: 0,
    server_out: '',
    update: update = 
        function (params) {
            this.world_id = typeof params.world_id === 'number' ? params.world_id : this.world_id;
            this.client_id = typeof params.client_id === 'number' ? params.client_id : this.client_id;
            this.server_out = typeof params.server_out !== undefined ? params.server_out : this.server_out;
        }
}

var send = function (data) {
    
    var url = '/api';
    data = JSON.stringify(data);
    $.post(globals.server_out+url, { json: data });
    console.log('socket.socket');
    console.log(socket.socket);
}

var admin, info, action;

admin = ( function () {

    var type = 'admin',
        create_agent,
        create_item,
        create_object,
        set_map,
        public_;
            
	create_agent = function (x, y, z, location) {
	
        var position;
    
        if (location === undefined) {
            location = 0;
        }
    
        if (x.constructor.name === 'Array') {
            if (x.length === 3) {
                x.splice(0,0,location);
            }
            position = x;
        } else {
            position = [location, x, y, z];
        }
    
		var data = { 
                cmd: 'create_agent',
                type: type,
                client_id: globals.client_id,
                world_id: globals.world_id,
                player_id: globals.player_id,
                position: position,
               }
        
        send(data);
	};
    
    create_object = function (x, y, z, location, object_type, template) {
        
        var position;
    
        if (location === undefined) {
            location = 0;
        }
    
        if (x.constructor.name === 'Array') {
            if (x.length === 3) {
                x.splice(0,0,location);
            }
            position = x;
        } else {
            position = [location, x, y, z];
        }
        
        var data = {
                type: type,
                cmd: 'create_item',
                object_type: object_type,
                template: template,
                client_id: globals.client_id,
                world_id: globals.world_id,
                position: position,
            }
        
        send(data);
    };
	
	set_map = function (x, y, z, value, debug) {
        
        if (debug) console.log(arguments);
        var position;
    
        if (value === undefined) {
            value = 1;
        }
    
        if (x.constructor.name === 'Array') {
            position = x;
            if (position.length < 4 && value !== undefined) {
                value = y;
                position.splice(0,0,value);
            }
        } else {
            position = [value, x, y, z];
        }

		var data = { 
                cmd: "set_map",
                type: type,
                client_id: globals.client_id,
                world_id: globals.world_id,
                position: position,
                value: value
               }
        if (debug) console.log(data);
        send(data);
	};
    
    public_ = {
                create_agent: create_agent,
                create_item: create_item,
                create_object: create_object,
                set_map: set_map
              };
              
    return public_;

}());


info = ( function () {
	
    var type = "info",
        consts = {},
        tileset, 
        map,
        agent,
        object,
        agents,
        objects,
        public_;
        
    consts = { type: type,
               client_id: globals.client_id,
               world_id: globals.world_id
             };
		
    tileset = function () {
        var data = $.extend(consts, { cmd: 'get_tiles', });
        send(data);
    };
        
	map = function (z) {

		var data = $.extend(consts, { cmd: "get_map", });
               
        if (z !== undefined) {
            data.z = z;
        }
		
        send(data);
	}

	agent = function (id) {

		var data = $.extend(consts, { cmd: "get_agent",
                                      agent_id: id, });
		send(data);	
	}

	object = function (id) {

		var data = $.extend(consts, { cmd: "get_object",
                                      object_id: id, });
		send(data);
	}
	
	agents = function() {
		
		var data = $.extend(consts, { cmd: "get_agent_list", });
		send(data);
	}

	objects = function() {
		
		var data = $.extend(consts, { cmd: "get_object_list", });
		send(data);
	}
	
    public_ = {
                tileset: tileset,
                map: map,
                agent: agent,
                object: object,
                agents: agents,
                objects: objects,
              };
              
    return public_;
    		
}());

action = ( function () {

    var type = 'agent',
        move,
        till,
        plant,
        harvest,
        public_;
		
	move = function (agent_id, dx, dy, dz) {

		var data = { 
                cmd: "move_0",
                type: type,
                client_id: globals.client_id,
                world_id: globals.world_id,
                agent_id: agent_id,
                dp: [dx, dy, dz]
               }
		
		send(data);
	}
    
    till = function (agent_id) {

        var data = {
                    type: type,
                    cmd: 'till_soil',
                    world_id: globals.world_id,
                    agent_id: agent_id,
                   }
                   
        send(data);
    },

    plant = function (agent_id) {

        var data = {
                    type: type,
                    cmd: 'plant_crop',
                    world_id: globals.world_id,
                    agent_id: agent_id,
                   }
                   
        send(data);
    },

    harvest = function (agent_id, crop_id) {
        
        //temporary
        if (crop_id === undefined) {
            crop_id = 1;
        }
        
        var data = {
                    type: type,
                    cmd: 'harvest_crop',
                    world_id: globals.world_id,
                    agent_id: agent_id,
                    crop_id: crop_id,
                   }
                   
        send(data);
    };
    
    public_ = {
                move: move,
                till: till,
                plant: plant,
                harvest: harvest,
              };
              
    return public_;
    
}());
